src/core/comm/coll.cc

```diff
@@ -0,0 +1,451 @@
+                 CollComm global_comm)
+{
+  log_coll.print("Alltoall: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#if defined(LEGATE_USE_GASNET)
+  return collAlltoallMPI(sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, global_comm);
+#else
+  return collAlltoallLocal(sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, global_comm);
+#endif
+}
+
+int collGather(const void* sendbuf,
+               int sendcount,
+               CollDataType sendtype,
+               void* recvbuf,
+               int recvcount,
+               CollDataType recvtype,
+               int root,
+               CollComm global_comm)
+{
+  log_coll.print("Gather: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#if defined(LEGATE_USE_GASNET)
+  return collGatherMPI(
+    sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, root, global_comm);
+#else
+  printf("Not implemented\n");
+  assert(0);
+#endif
+}
+
+int collAllgather(const void* sendbuf,
+                  int sendcount,
+                  CollDataType sendtype,
+                  void* recvbuf,
+                  int recvcount,
+                  CollDataType recvtype,
+                  CollComm global_comm)
+{
+  log_coll.print("Allgather: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#if defined(LEGATE_USE_GASNET)
+  return collAllgatherMPI(sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, global_comm);
+#else
+  return collAllgatherLocal(
+    sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, global_comm);
+#endif
+}
+
+int collBcast(void* buf, int count, CollDataType type, int root, CollComm global_comm)
+{
+  log_coll.print("Bcast: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#if defined(LEGATE_USE_GASNET)
+  return collBcast(buf, count, type, root, global_comm);
+#else
+  printf("Not implemented\n");
+  assert(0);
+#endif
+}
+
+// called from main thread
+int collInit(int argc, char* argv[])
+{
+  current_unique_id = 0;
+#if defined(LEGATE_USE_GASNET)
+  int provided, res, init_flag = 0;
+  MPI_Initialized(&init_flag);
+  if (!init_flag) {
+    res = MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+    assert(res == MPI_SUCCESS);
+  } else {
+    log_coll.print(
```

Comment:
do we actually hit this case? I feel we should just abort in this case, as it clearly indicates that this code isn't used as intended.

Reply:
Yes, we do. GASNet may init MPI, which is before the `collInit` is called.

Reply:
Your use of "may" confuses me. Does GASNet always initialize MPI or can it sometimes not initialize MPI before this function is called? If it's the former, we don't need this branch. Otherwise, I think we should understand when that can happen, and eliminate the uncertainty if possible.

Reply:
Actually, I am not sure about it. I have been used ibv conduit, and GASNet indeed initializes MPI. So I add this warning just to let people know if they see errors, it may be because of the MPI init.

Reply:
Is there a way to check if the MPI was initialized with `MPI_THREAD_MULTIPLE`? This warning message doesn't seem very useful, as the users have no way to fix it. Users will almost certainly use the install script that we provide, so if GASNet didn't initialize MPI with `MPI_THREAD_MULTIPLE`, it's actually our fault. And if we make sure GASNet always initializes MPI correctly, this warning message is unnecessary. I guess this is yet another issue that we won't have with UCX, right?

Reply:
Ok, so that sounds like a real problem, and once we fixed that, we no longer need to have this check, as MPI will always get initialized correctly by Legate. Let me figure out where to add the init call.

Reply:
I found the MPI function that can check if MPI_THREAD_MULTIPLE is used, and I have updated the code accordingly.